From 9f53930476936329eeb40b010dab8925227da061 Mon Sep 17 00:00:00 2001
From: Johannes Schindelin <johannes.schindelin@gmx.de>
Date: Thu, 5 Oct 2017 14:41:59 +0200
Subject: [PATCH 56/N] fixup! Avoid comparing `this` to `NULL`

Turns out that these comparisons are actually essential: otherwise e.g.
`kill <pid>` will simply segfault if `<pid>` does not refer to an MSYS2
process (see https://github.com/git-for-windows/git/issues/1316 for a
real-world bug report).

We *have* to compile this code with -Wno-error=nonnull-compare.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>
---
 winsup/cygwin/fhandler_dsp.cc |  4 ++--
 winsup/cygwin/pinfo.cc        | 18 +++++++++---------
 winsup/cygwin/signal.cc       |  2 +-
 3 files changed, 12 insertions(+), 12 deletions(-)

diff --git a/winsup/cygwin/fhandler_dsp.cc b/winsup/cygwin/fhandler_dsp.cc
index 5a6592f..5ae3309 100644
--- a/winsup/cygwin/fhandler_dsp.cc
+++ b/winsup/cygwin/fhandler_dsp.cc
@@ -500,7 +500,7 @@ fhandler_dev_dsp::Audio_out::buf_info (audio_buf_info *p,
 				       int rate, int bits, int channels)
 {
   p->fragstotal = MAX_BLOCKS;
-  if (dev_)
+  if (this && dev_)
     {
       /* If the device is running we use the internal values,
 	 possibly set from the wave file. */
@@ -957,7 +957,7 @@ fhandler_dev_dsp::Audio_in::buf_info (audio_buf_info *p,
 {
   p->fragstotal = MAX_BLOCKS;
   p->fragsize = blockSize (rate, bits, channels);
-  if (dev_)
+  if (this && dev_)
     {
       p->fragments = Qisr2app_->query ();
       if (pHdr_ != NULL)
diff --git a/winsup/cygwin/pinfo.cc b/winsup/cygwin/pinfo.cc
index f7893a2..8265052 100644
--- a/winsup/cygwin/pinfo.cc
+++ b/winsup/cygwin/pinfo.cc
@@ -529,7 +529,7 @@ _pinfo::set_ctty (fhandler_termios *fh, int flags)
 bool __reg1
 _pinfo::exists ()
 {
-  return process_state && !(process_state & (PID_EXITED | PID_REAPED | PID_EXECED));
+  return this && process_state && !(process_state & (PID_EXITED | PID_REAPED | PID_EXECED));
 }
 
 bool
@@ -720,7 +720,7 @@ _pinfo::commune_request (__uint32_t code, ...)
   res.s = NULL;
   res.n = 0;
 
-  if (!pid)
+  if (!this || !pid)
     {
       set_errno (ESRCH);
       goto err;
@@ -819,7 +819,7 @@ out:
 fhandler_pipe *
 _pinfo::pipe_fhandler (int64_t unique_id, size_t &n)
 {
-  if (!pid)
+  if (!this || !pid)
     return NULL;
   if (pid == myself->pid)
     return NULL;
@@ -832,7 +832,7 @@ char *
 _pinfo::fd (int fd, size_t &n)
 {
   char *s;
-  if (!pid)
+  if (!this || !pid)
     return NULL;
   if (pid != myself->pid)
     {
@@ -856,7 +856,7 @@ char *
 _pinfo::fds (size_t &n)
 {
   char *s;
-  if (!pid)
+  if (!this || !pid)
     return NULL;
   if (pid != myself->pid)
     {
@@ -884,7 +884,7 @@ char *
 _pinfo::root (size_t& n)
 {
   char *s;
-  if (!pid)
+  if (!this || !pid)
     return NULL;
   if (pid != myself->pid && !ISSTATE (this, PID_NOTCYGWIN))
     {
@@ -930,7 +930,7 @@ char *
 _pinfo::cwd (size_t& n)
 {
   char *s = NULL;
-  if (!pid)
+  if (!this || !pid)
     return NULL;
   if (ISSTATE (this, PID_NOTCYGWIN))
     {
@@ -976,7 +976,7 @@ char *
 _pinfo::cmdline (size_t& n)
 {
   char *s = NULL;
-  if (!pid)
+  if (!this || !pid)
     return NULL;
   if (ISSTATE (this, PID_NOTCYGWIN))
     {
@@ -1036,7 +1036,7 @@ char *
 _pinfo::environ (size_t& n)
 {
   char **env = NULL;
-  if (!pid)
+  if (!this || !pid)
     return NULL;
   if (ISSTATE (this, PID_NOTCYGWIN))
     {
diff --git a/winsup/cygwin/signal.cc b/winsup/cygwin/signal.cc
index 93be358..fbd2b24 100644
--- a/winsup/cygwin/signal.cc
+++ b/winsup/cygwin/signal.cc
@@ -260,7 +260,7 @@ _pinfo::kill (siginfo_t& si)
 	}
       this_pid = pid;
     }
-  else if (process_state == PID_EXITED)
+  else if (this && process_state == PID_EXITED)
     {
       this_process_state = process_state;
       this_pid = pid;
-- 
2.9.0

